Prevent open requests from leaking when HTTP2.request or HTTP2. stream_request_body return an error

mezz opened the pull request:
Hello, and thank you for the great library!

## Issue Description

HTTP2 leaks open requests in some cases when there are errors raised in `HTTP2.request/5`.

## Background

I ran into this when hitting `:exceeds_window_size` and `:too_many_concurrent_requests` errors in production.
We get some of these errors during the normal operation of our service, and each one leaks a request. When this happens, we retry our request on a different HTTP2 connection, so the leak was not noticeable for a long time.

## Impact of the issue

Unfortunately we can eventually end up with enough leaked requests so that every HTTP2 request will error with `:too_many_concurrent_requests`. This leaves the HTTP2 connection in a bad state where it will no longer work.

## PR Changes

This PR introduces some logic to clean up these errored requests, and adds a few checks to the tests to confirm the fix.

ericmj commented on lib/mint/http2.ex:
```diff
@@ -491,24 +491,16 @@ defmodule Mint.HTTP2 do
+      :throw, {:mint, conn, reason} ->
+        stream = Map.fetch!(conn.streams, stream_id)
+        conn = delete_stream(conn, stream)
+        {:error, conn, reason}
```
We need to return the original `conn` instead of calling `delete_stream` because there is other state on the `conn` that needs to be reverted, such as window_size and number of open streams.

This code will also be run when `send!` fails but I guess that's fine since the connection shouldn't be used after a transport error anyway.

mezz replied:
Thanks for the review.

`delete_stream` will revert the number of open streams, I have added tests for it so that part should be good.

However `conn.window_size` is updated fairly deep, in `encode_data_chunk`. I will add some test checks for the window size and see if I can come up with a good strategy to reliably revert it.

ericmj replied:
You should be able to return the unmodified original conn that is passed in as a function argument.

mezz replied:
Oh thanks for the clarification, I understand what you mean now. I was working on refactoring to have the connection track the stream only at the very end, but I think your approach will be safer and easier so I'll try that out with new tests on Monday.

ericmj commented:
Thanks for finding the issue and sending a PR with a fix! I think we need to do the same fix in `stream_request_body`.

mezz commented:
I have:
 * updated this to return the original `conn` instead of calling `delete_stream`
 * added the same fix for `stream_request_body`
 * added more tests

To test that transport errors are handled here, I created a `TestTransportSendTimeout` module which is the same as the `SSL` transport module except it always times out on successful `send`s. The way I use it in the tests seems a bit less clean than would be ideal, but there is currently no way to use a transport other than `TCP` and `SSL` and I didn't want to go down that rabbit hole in this PR. Let me know if that works for you or if there's a better approach you'd like to use for these.

ericmj commented on lib/mint/http2.ex:
```diff
@@ -482,35 +482,25 @@ defmodule Mint.HTTP2 do
+      :throw, {:mint, _conn, reason} ->
+        # The stream is invalid and "_conn" may be tracking it, so we return the original connection instead.
+        {:error, original_conn, reason}
+    end
```
The `conn` that is in scope here is the original `conn` from the function arguments. So you don't have to do `original_conn = conn` and can revert the `try do`.

mezz replied:
Sounds good. I have updated these to use the same `catch` structure that the code had before, but updated to use the original `conn`.

ericmj commented:
Thank you! 💜
